File: tree_sort.cc
/*
 * Nov 23, 2018
 * Tree sort..
 */
#include <iostream>
#include <vector>

using namespace std;

struct Node {
  int value;
  Node* left;
  Node* right;
};

Node* new_node(int val) {
  Node* node = new Node();
  node->value = val;
  node->left = nullptr;
  node->right = nullptr;
  return node;
}

Node* plant(int val, Node* node) {
  if (node == nullptr) {
    node = new_node(val);
    return node;
  } else {
    if (val < node->value)
      node->left = plant(val, node->left);
    else
      node->right = plant(val, node->right);
  }
  return node;
}

void harvest(Node* root, vector<int>& sorted) {
  if (root == nullptr) return;
  harvest(root->left, sorted);
  sorted.push_back(root->value);
  harvest(root->right, sorted);
}

vector<int> tree_sort(vector<int>& arr) {
  vector<int> leaves{};
  if (arr.size() == 0) return leaves;
  Node* seed = nullptr;
  Node* root = plant(arr[0], seed);
  for (int i = 1; i < arr.size(); ++i) {
    plant(arr[i], root);
  }

  harvest(root, leaves);
  return leaves;
}

void print_vec(vector<int>& vec) {
  for (auto v : vec) {
    cout << v << " ";
  }
  cout << endl;
  return;
}

int main() {
  vector<int> arr{2037, -1, 5, 4, 100, 4};
  print_vec(arr);
  auto sorted_arr = tree_sort(arr);
  print_vec(sorted_arr);

  return 0;
}

File: tree-sort.js
/*
 * Nov 23, 2018
 * Tree sort in JavaScript
 */

let Node = function(val) {
  this.value = val;
  this.left = null;
  this.right = null;
};

let plant = function(key, node) {
  if (node === null) {
    node = new Node(key);
    return node;
  }
  if (key < node.value) node.left = plant(key, node.left);
  else node.right = plant(key, node.right);

  return node;
};

let harvest = function(node, sorted_arr = []) {
  if (node === null) return;
  harvest(node.left, sorted_arr);
  sorted_arr.push(node.value);
  harvest(node.right, sorted_arr);
};

let treeSort = function(arr) {
  if (arr.length === 0) return [];
  let seed = null;
  let root = plant(arr[0], seed);
  for (let i = 1; i < arr.length; i++) {
    plant(arr[i], root);
  }
  let sorted = []
  harvest(root, sorted);
  return sorted;
};

// let sorted = treeSort([2037, -1, 5, 4, 100, 4]);
// console.log(sorted);
console.log(treeSort([2037, -1, 5, 4, 100, 4]));
